## Handwritten Digits Classification
### Using Multilayer Perceptron

In [5]:
import keras
from keras.datasets import mnist


In [7]:
# load the dataset
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

#about 
print('Training set shape: ',train_images.shape)
print('Length of training set: ', len(train_labels))

print('Testing set shape: ', test_images.shape)
print('Length of testing set: ', len(test_labels))

Training set shape:  (60000, 28, 28)
Length of training set:  60000
Testing set shape:  (10000, 28, 28)
Length of testing set:  10000


### Workflow
- Present the neural network with the training data, `train_images` and `train_labels`.
> The network will then learn to associate images and labels.
- Finally we ask the network to produce predictions for `test_images`, and we will verify these predictions by matching them with the labels from `test_labels`.

The 'layer' is a data-processing module which can also be percieved as a 'filter' for data.

- Our network consists of a sequence of two `Dense` layers, which are densly connectedneural layers( fully-connected ).
- The 2nd layer is a 10 way 'softmax' layer, which means it will return an array of 10 probability scores.
- Each score will be the probability that the current digit belongs to one of the 10 digit classes.

We also need to define:
- A `loss function`
- An `optimizer`
- `Metrics`

In [11]:
from keras import models, layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28*28,)))
network.add(layers.Dense(10, activation='softmax'))

network.compile(optimizer = 'rmsprop',
               loss = 'categorical_crossentropy',
               metrics = ['accuracy'])

### Preprocessing and Reshaping
We scale our data so that all values are in the `[0, 1]` interval.

Previously, training images were stored in an array of shape `(60000, 28, 28)` of type `uint8` with values in `[0, 255]` interval.
We transform it into a `float32` array of shape `(60000, 28*28)` with values between 0 and 1.

In [12]:
train_images = train_images.reshape((60000, 28*28))
train_images = train_images.astype('float32')/255

test_images = test_images.reshape((10000, 28*28))
test_images = test_images.astype('float32')/255

We also categorically encode the labels:

In [14]:
from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

### Training the network
We call the `fit` method to train the network i.e, we 'fit' the model to its training data.

During training two quantities are displayed:
- 'Loss' of the network over the training data
- 'Accuracy' of the network over the training data

In [17]:
network.fit(train_images, train_labels, epochs=6, batch_size=128)

Epoch 1/6
469/469 [==============================] - 4s 9ms/step - loss: 0.0290 - accuracy: 0.9912
Epoch 2/6
469/469 [==============================] - 4s 9ms/step - loss: 0.0216 - accuracy: 0.9939
Epoch 3/6
469/469 [==============================] - 4s 9ms/step - loss: 0.0169 - accuracy: 0.9953
Epoch 4/6
469/469 [==============================] - 4s 9ms/step - loss: 0.0128 - accuracy: 0.9964
Epoch 5/6
469/469 [==============================] - 4s 9ms/step - loss: 0.0095 - accuracy: 0.9977
Epoch 6/6
469/469 [==============================] - 6s 12ms/step - loss: 0.0071 - accuracy: 0.9984


We reached an accuracy of `0.9984( i.e, 99.8% )` on the training data.
Now check the model's performance on testing data too.

In [18]:
test_loss, test_acc = network.evaluate(test_images, test_labels)

313/313 [==============================] - 2s 5ms/step - loss: 0.0689 - accuracy: 0.9809


Our test set accuracy turns out to be `98.09%`, which is slightly less than the training set accuracy.